DM-21263: Implement packaging of spectrograph output

parejkoj opened the pull request.

timj commented on python/lsst/ts/FiberSpectrograph/dataManager.py:
```diff
@@ -0,0 +1,130 @@
+
+    def make_wavelength_hdu(self, data):
+        """Return the wavelength HDU build from SpectrographData."""
+        hdu = astropy.io.fits.ImageHDU(data=data.wavelength)
```
Is there any way we can put the units of this wavelength in metadata somewhere? If this is the wavelength of every pixel in the primary HDU then it could/should be stored as a standard WCS associated with the data itself (using the -TAB format if needs be; Greisen et al 2006 A&A, 446, 747).

parejkoj replied:
After having spent a few days digging into this, it looks like neither astropy nor specutils currently supports the spectral `-TAB` WCS format. You can make a `gwcs` tablular spectral wcs, but it doesn't support writing it to FITS as a G. 2006 style FITS WCS.

I think I might file another ticket and defer making a FITS WCS here until one of the above properly supports it.

parejkoj replied:
See this PR I made to specutils: https://github.com/astropy/specutils/pull/530
and the latest discussion on this astropy Issue: https://github.com/astropy/astropy/issues/7883

parejkoj replied:
I've changed to outputting the wavelength data as a BinTable HDU with unit information.

Given the fact that astropy and specutils don't fully support this type of WCS, I don't have a good way to verify that I've done it correctly if I try to do it with astshim. I can make the AST FrameDict, but I don't know how to turn that into a Greisen 2006-style WCS:

```python
pixelFrame = astshim.Frame(1, "Domain=PIXELS");
wavelengthFrame = astshim.Frame(1, "Domain=WAVELENGTH");
wavelength = np.arange(300, 1000, 1000)
lutMap = astshim.LutMap(wave, 0, 1)
frameDict = astshim.FrameDict(pixelFrame)
frameDict.addFrame("PIXELS", lutMap, wavelengthFrame)
```

timj replied:
Look up the TabOK parameter in AST for when you want to generate the FitsChan from the WCS.

timj commented on tests/test_dataManager.py:
```diff
@@ -0,0 +1,120 @@
+                                     n_pixels=self.n_pixels)
+
+    def check_header(self, header):
+        self.assertEqual(header['TELESCOP'], self.telescope)
```
If the header compares as a dict then you can set up the output dict once and then use `self.assertEqual(header, self.reference_header)`  Depends whether you are worried about extra headers turning up that you aren't expecting.

parejkoj replied:
It doesn't look like dict-like comparison works for astropy Headers: the header is being compared as a big FITS character block instead of a dict.

parejkoj replied:
Also, we can't do a direct comparison, because there will be added keys in the final output FITS header (e.g. 'SIMPLE').

parejkoj replied:
I created a dict in the class and iterated over its `items()`.

timj commented on python/lsst/ts/FiberSpectrograph/dataManager.py:
```diff
@@ -0,0 +1,130 @@
+
+@dataclasses.dataclass
+class SpectrographData:
+    """Class to hold data and metadata from a fiber spectrograph.
```
Does this include all the things that you set which controls data acquisition?

parejkoj replied:
I'm not sure what you mean. The only things that matter are the number of pixels (fixed for a given spectrograph) and the duration of the exposure. I've not implemented sub-frame readout, and I don't think it would be necessary or desirable (readout time is a handful of milliseconds, so subframes would just make downstream processing harder).

There was talk of doing an "exposure loop" (take 10 exposures of the same duration), but whether that is implemented at the CSC level or in the controller object hasn't been determined, so we don't know what metadata we'd need to write out for it.

parejkoj commented on python/lsst/ts/FiberSpectrograph/dataManager.py:
```diff
@@ -0,0 +1,130 @@
+            return None
+
+    def make_fits_header(self, data):
+        """Return a FITS header built from SpectrographData."""
```
Said python dict would already have FITS keys baked into it though; why else would you name a dictionary key `INSTRUME` or `EXPTIME` instead of `instrument` and `exposure_time`? That seems like a distinction without a difference.

timj replied:
If you start dealing with comment strings and units then it clearly makes sense to work with FITS objects here (or PropertyList if you were in DM).

parejkoj commented on python/lsst/ts/FiberSpectrograph/dataManager.py:
```diff
@@ -0,0 +1,130 @@
+        hdr['SOURCE'] = data.source
+        hdr['TEMP_SET'] = data.temperature_setpoint
+        hdr['CCDTEMP'] = data.temperature
+        return hdr
```
:+1: on ORIGIN.

The units of the flux data are in instrumental counts: we have to do an external spectrophotometry if we want proper units on them.

parejkoj commented on python/lsst/ts/FiberSpectrograph/dataManager.py:
```diff
@@ -0,0 +1,130 @@
+    """The light source that was measured (see the XML `expose.source` field).
+    """
+    temperature: float
+    """The internal spectrograph temperature."""
```
Although that is a good idea, it doesn't much us when we write a FITS file, since it doesn't look like there is a way to write the quantity objects to a header (instead of a binary table). I've filed an astropy issue about this:

https://github.com/astropy/astropy/issues/9332

timj replied:
In the short term you could put the unit in a comment yourself -- comments are not currently being used and your code to build the fits header could check for quantities and behave properly.

parejkoj replied:
I'd rather not add custom unit handling here: we might as well let astropy handle it directly, if/when they implement support for the above issue.

parejkoj commented:
I've updated the code based on your comments. Please have another look.

As part of that, I've added a format version number to the output, for when we eventually get a spectral WCS output (or other changes).
